### PR TITLE
Added Type Hints for modeling_tf_encoder_decoder.py

### PR DESCRIPTION
**Pull request by Batese2001**
# What does this PR do?

This pull request adds type hints for modeling_tf_encoder_decoder.py as outlined in Issue #16059 while not being on my main branch

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@Rocketknight1

### CODE REVIEW

**Comment by Batese2001** [comment]
@Rocketknight1 I think this is ready, though I am getting an odd error saying "module 'tensorflow' has no attribute 'tensor'" which  I am not sure how to resolve. Thanks and let me know if there is anything I need to fix!

File: src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py [hunk 1]
```diff
@@ -510,22 +517,22 @@ def from_encoder_decoder_pretrained(
...
-        output_attentions=None,
-        output_hidden_states=None,
-        return_dict=None,
-        training=False,
+        input_ids: Optional[TFModelInputType] = None,
+        attention_mask: Optional[Union[np.ndarray, tf.Tensor]] = None,
+        decoder_input_ids: Optional[Union[np.ndarray, tf.Tensor]] = None,
+        decoder_attention_mask: Optional[Union[np.ndarray, tf.Tensor]] = None,
+        encoder_outputs: Optional[Union[np.ndarray, tf.Tensor]] = None,
+        past_key_values: Optional[Tuple[Tuple[tf.tensor]]] = None,
```
Comment (Rocketknight1) [inline src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py:525]:
```suggestion
        past_key_values: Optional[Tuple[Tuple[tf.Tensor]]] = None,
```

**Comment by Rocketknight1** [comment]
@Batese2001 The issue was that one of your hints was `tf.tensor` instead of `tf.Tensor` - it was case-sensitive! I just changed it there, let's see if that fixes the tests.

**Comment by Rocketknight1** [comment]
Tests look good - that Torch error is totally unrelated to this PR. Are you happy for me to merge the PR at this point?

**Comment by Batese2001** [comment]
Wonderful! If you think it is ready, then I am all for merging! Thank you for your help

**Review by Rocketknight1 — APPROVED** [review]

**Comment by Rocketknight1** [comment]
Merged, and thanks for your help!

**Pull request merged**